docs/adr/0006-filtering-system-and-parameters.md

```diff
@@ -0,0 +1,253 @@
+
+We know `author` is a property of `Book`, that represents a Resource. So it can be filtered by:
+
+- IRI
+- uid
+
+We should therefore call both of these filters for each query parameter matched: 
+
+- IriFilter (will do nothing if the value is not an IRI)
+- UuidFilter
+
+With that in mind, an `or` filter would call a bunch of filters specifying the logic operation to execute. 
+
+### Query parameter
+
+The above shows that a query parameter **key**, which is a `string` may lead to multiple filters being called. This same can represent one or multiple values, and for a same **key** we can handle multiple types of data.
+Also, if someone wants to implement the [loopback API](https://loopback.io/doc/en/lb2/Fields-filter.html) `?filter[fields][vin]=false` the link between the query parameter, the filter and the value gets more complex. 
+
+We need a way to instruct the program to parse query parameters and produce a link between filters, values and some context (property, logical operation, type etc.). The same system could be used to determine the **type** a **filter** must have to pilot query parameter validation and the JSON Schema. 
+
+## Considered Options
+
+Let's define a new Attribute `Parameter` that holds informations (filters, context, schema) tight to a parameter `key`. 
+
+```php
+namespace ApiPlatform\Metadata;
+
+use ApiPlatform\OpenApi;
+
+final class Parameter {
+    public string $key;
+    public \ArrayObject schema;
+    public array $context;
+    public OpenApi\Parameter $openApi;
+    public string|callable provider(): Operation;
+    // filter service id
+    public string $filter;
+}
+```
+
+By default applied to a class, the `Parameter` would apply on every operations, or it could be specified on a single operation: 
+
+```php
+use ApiPlatform\Metadata\GetCollection;
+use ApiPlatform\Doctrine\Common\AndParameter;
+
+#[GetCollection(parameters: ['and' => new AndParameter])]
+#[AndParameter('and')]
+class Book {}
+```
+
+API Platform will continue to provide parsed query parameters and set an `_api_query_parameters` Request attribute, in the end the filter may or may not use it:
+
+```php
+$queryString = RequestParser::getQueryString($request);
+$request->attributes->set('_api_query_parameters', $queryString ? RequestParser::parseRequestParams($queryString) : []);
+```
+
+On top of that we will provide an additional `_api_header_parameters` as we would like to introduce a `QueryParameter` and an `HeaderParameter`. 
+
+### Parameter Provider
+
+During the `Provider` phase (`RequestEvent::REQUEST`), we could use a `ParameterProvider`:
+
+```php
+/**
+ * Optionnaly transforms request parameters and provides modification to the current Operation.
+ *
+ * @implements ProviderInterface<HttpOperation>
+ */
+interface ParameterProvider extends ProviderInterface {
+    public function provider(Operation $operation, array $uriVariables = [], array $context = []): HttpOperation;
```

Comment:
Why not `Operation` instead of `HttpOperation`?

Reply:
indeed I used HttpOperation inside the implementation